src/datamodel/builder/UMessage.cpp

```diff
@@ -10,3 +10,245 @@
+		if (priority < v1::UPriority::UPRIORITY_CS4) {
+			throw std::out_of_range(
+			    "Priority value is out of range, for req/resp should be > CS4");
+		}
+	}
+
+	attributes_.set_priority(priority);
+
+	return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withTtl(std::chrono::milliseconds ttl) {
+	if ((ttl.count() <= 0) ||
+	    (ttl.count() > std::numeric_limits<uint32_t>::max())) {
+		throw std::out_of_range("TTL value is out of range");
+	}
+
+	attributes_.set_ttl(static_cast<uint32_t>(ttl.count()));
+
+	return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withToken(const std::string& token) {
+	if ((attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_REQUEST)) {
+		throw std::domain_error("Token can only be set on a request message");
+	}
+
+	attributes_.set_token(token);
+
+	return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withPermissionLevel(uint32_t level) {
+	if ((attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_REQUEST)) {
+		throw std::domain_error(
+		    "Permission level can only be set on a request message");
+	}
+
+	attributes_.set_permission_level(level);
+
+	return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withCommStatus(v1::UCode code) {
+	if (attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_RESPONSE) {
+		throw std::domain_error(
+		    "CommStatus can only be set on a response message");
+	}
+
+	if (!(v1::UCode_IsValid(code))) {
+		throw std::out_of_range("UCode is out of range" + std::to_string(code));
+	}
+
+	if (code != v1::UCode::OK) {
+		attributes_.set_commstatus(code);
+	}
+
+	return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withPayloadFormat(v1::UPayloadFormat format) {
+	setPayloadFormat(format);
+
+	return *this;
+}
+
+v1::UMessage UMessageBuilder::build() {
+	v1::UMessage message;
+	if (expectedPayloadFormat_.has_value()) {
+		throw UnexpectedFormat(
+		    "Tried to build with no payload when a payload format has been set "
+		    "using withPayloadFormat()");
+	}
+
+	*attributes_.mutable_id() = uuidBuilder_->getBuilder().build();
+	*message.mutable_attributes() = attributes_;
+
+	return message;
+}
+
+v1::UMessage UMessageBuilder::build(builder::Payload&& payload) const {
+	v1::UMessage message;
+
+	*(message.mutable_attributes()->mutable_id()) =
+	    uuidBuilder_->getBuilder().build();
+	*message.mutable_attributes() = attributes_;
+	auto [payloadData, payloadFormat] = std::move(payload).buildMove();
+	if (expectedPayloadFormat_.has_value()) {
+		if (payloadFormat != expectedPayloadFormat_) {
+			throw UnexpectedFormat(
+			    "Payload format does not match the expected format");
+		}
+	}
+	*message.mutable_payload() = std::move(payloadData);
+	message.mutable_attributes()->set_payload_format(payloadFormat);
+
+	return message;
+}
+
+UMessageBuilder::UMessageBuilder(v1::UMessageType msgType, v1::UUri&& source,
+                                 std::optional<v1::UUri>&& sink,
+                                 std::optional<v1::UUID>&& request_id) {
```

Comment:
`uuidBuilder_` is not initialized
```suggestion
UMessageBuilder::UMessageBuilder(v1::UMessageType msgType, v1::UUri&& source,
                                 std::optional<v1::UUri>&& sink,
                                 std::optional<v1::UUID>&& request_id) 
                                 : uuidBuilder_(UuidBuilder::getBuilder()) {
```